tests/legacy-cli/e2e/tests/commands/add/secure-registry.ts

```diff
@@ -4,9 +4,12 @@ import { createNpmConfigForAuthentication } from '../../../utils/registry';
 import { expectToFail } from '../../../utils/utils';
 
 export default async function () {
-  // The environment variable has priority over the .npmrc
-  const originalRegistryVariable = process.env['NPM_CONFIG_REGISTRY'];
-  delete process.env['NPM_CONFIG_REGISTRY'];
+  let originalRegistryVariable;
```

Comment:
Why are the following changes needed?

Reply:
When debugging the tests that were failing after my changes, I found that at least two of the tests following this pattern  - *secure-registry* and *registry-option* - ended up setting `NPM_CONFIG_REGISTRY` to a string value of "undefined".  This caused subsequent tests to fail now that the environment variables are being read in.  I figured I should update all three occurrences of this pattern to be thorough.  Only after making these changes was I able to get the entire test suite to pass.

Reply:
@alan-agius4 if you look at my latest test failure, I dumped the environment variables after all the setup scripts have completed (search for `env-vars-post-500`) and right before the *secure-registry* script tries to save the `NPM_CONFIG_REGISTRY` value.  In both cases the value doesn't exist hence the `undefined` string value is set when being put back when the test completes.  

Do you know why the registry value isn't set in the `setRegistry` function when `ci` is set to true?  That might be the root of the issue here from what I'm seeing.

Reply:
On CI we can use the `npm config` directly. Setting `NPM_CONFIG_REGISTRY` to `undefined` shouldn't be a problem at all. Let me try to get to the bottom of this.

Reply:
Here's the stack 

```
[error] TypeError [ERR_INVALID_URL]: Invalid URL: undefined/undefined/@angular%2fcli
    at onParseError (internal/url.js:259:9)
    at new URL (internal/url.js:335:5)
    at regFetch (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/npm-registry-fetch/index.js:48:5)
    at RegistryFetcher.packument (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/pacote/lib/registry.js:86:15)
    at RegistryFetcher.manifest (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/pacote/lib/registry.js:117:17)
    at Object.manifest (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/pacote/lib/index.js:16:45)
    at Object.fetchPackageManifest (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/@angular/cli/utilities/package-metadata.js:215:35)
    at UpdateCommand.checkCLILatestVersion (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/@angular/cli/commands/update-impl.js:658:60)
    at UpdateCommand.run (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/@angular/cli/commands/update-impl.js:218:49)
    at async UpdateCommand.validateAndRun (/mnt/ramdisk/e2e-main/subdirectory/subdirectory-test-project/node_modules/@angular/cli/models/command.js:136:22)

```

Reply:
Note: the tests which is failing `tests/misc/update-git-clean-subdirectory.ts` ran prior to `tests/update/update-secure-registry.ts`.

```diff
setup/010-local-publish.ts
setup/100-global-cli.ts
setup/200-create-tmp-dir.ts
setup/300-log-environment.ts
setup/500-create-project.ts
tests/basic/e2e.ts
tests/basic/serve.ts
tests/build/barrel-file.ts
tests/build/bundle-budgets.ts
tests/build/differential-cache.ts
tests/build/json.ts
tests/build/no-sourcemap.ts
tests/build/prod-build.ts
tests/build/rebuild-replacements.ts
tests/build/styles/imports.ts
tests/build/styles/preset-env.ts
tests/build/ts-paths.ts
tests/commands/add/add.ts
tests/commands/add/registry-option.ts
tests/commands/config/config-global.ts
tests/commands/help/help-json.ts
tests/commands/unknown-configuration.ts
tests/generate/component/component-duplicate.ts
tests/generate/component/component-module-export.ts
tests/generate/directive/directive-basic.ts
tests/generate/generate-error.ts
tests/generate/help-output-no-duplicates.ts
tests/generate/library/library-consumption-ve.ts
tests/generate/pipe/pipe-basic.ts
tests/generate/service/service-basic.ts
tests/i18n/ivy-localize-dl-json.ts
tests/i18n/ivy-localize-hashes.ts
tests/i18n/ivy-localize-sourcelocale.ts
tests/misc/coverage.ts
tests/misc/forwardref-es2015.ts
tests/misc/minimal-config.ts
tests/misc/npm-audit.ts
tests/misc/target-default-configuration.ts
- tests/misc/update-git-clean-subdirectory.ts
tests/schematics_cli/basic.ts
tests/test/test-scripts.ts
+ tests/update/update-secure-registry.ts
```

Reply:
Ill continue trying to get to the bottom of this later on today.

Reply:
So what is happening is that Node.JS will serialize undefined to `'undefined'`. This means that the check in https://github.com/angular/angular-cli/pull/21297/files#diff-1fe35077f343d441cac1946e85d88adf8808fa6b3a0898e85209cc7143e6b9c0R151 should be changed since to reset a variable one would need to set an empty string.

Reply:
Same result.  A string of `undefined` will still result in a truthy value so it will still set the option from the environment variable.  I can either update my code in *package-metadata* to look for a string value of `undefined` or do the changes I proposed earlier in those tests to make sure they don't set an undefined value to `undefined` when leaving the test.  In the second test suite failure it sets the env var to `undefined` when leaving *secure-registry* which then causes *update-git-clean* to fail.

![image](https://user-images.githubusercontent.com/575062/126330225-d90535c6-c827-437f-b669-4bccc5b13e46.png)

Those are the only two options that I see, but I'm open to other suggestions.

Reply:
Let just update the spec files, since checking for `"undefined"` isn't correct.

Reply:
Sounds good.